Source/DafnyCore/AST/Grammar/ParseErrors.cs

```diff
@@ -407,18 +510,564 @@ declared identifier.
+predicate name and the opening parenthesis of the formals.
+The type may be either `nat` or `ORDINAL`.
+This special formal is not permitted in a regular (non-extreme) predicate.
+", Remove(true)); // TODO - would like to have option to add 'least' or 'greatest', but that takes a different token location
+
+    Add(ErrorId.p_no_by_method_in_twostate,
+    @"
+Two state functions and predicates are always ghost and do not have a compiled representation.
+Such functions use values from two different program (heap) states, which is not 
+something that can be implemented (at least with any degree of good performance) in conventional programming languages.
+
+Because there is no feasible compiled verion of a two-state function, it cannot have a `by method` block (which is always compiled).
+", Remove(false, "remove by method block"));
+
+    Add(ErrorId.p_no_by_method_in_extreme_predicate,
+    @"
+Least and greatest predicates are always ghost and do not have a compiled representation, 
+so it makes no sense to have a `by method` alternative implementation.
+", Remove(false, "remove by method block"));
+
+    Add(ErrorId.p_no_by_method_for_ghost_function,
+    @"
+`by method` constructs combine a ghost function (or predicate) with a non-ghost method.
+The two parts compute the same result, and are verified to do so.
+Uses of the function are verified using the function body, but the method body is used when the function is compiled.
+
+Thus the function part is always implicitly `ghost` and may not be explicitly declared `ghost`.
+", Remove(true));
+
+    Add(ErrorId.p_twostate_and_extreme_are_always_ghost,
+    @"
+The `twostate`, `least`, and `greatest` functions and predicates are always ghost and cannot be compiled, so they cannot be
+declared as a `function method` (v3 only).
+", Remove(true));
+
+    Add(ErrorId.p_old_ghost_syntax,
+    @"
+This error only occurs when the `experimentalPredicateAlwaysGhost` option is chosen.
+It indicates that `predicates` are always ghost and cannot be declared with the (Dafny 3) syntax `predicate method`.
+- If you intend to predicate to be ghost, remove `method`.
+- If you intend the predicate to be non-ghost, you either cannot use `experimentalPredicateAlwaysGhost` or you should use `function` with a `bool` return type instead of `predicate`
+", Remove(true));
+
+    Add(ErrorId.p_deprecating_predicate_method,
+    @"
+    From Dafny 4 on, the phrases `function method` and `predicate method` are no
+    longer accepted. Use `function` for compiled, non-ghost functions and
+    `ghost function` for non-compiled, ghost functions, and similarly for predicates.
+    See [the documentation here](https://dafny.org/latest/DafnyRef/DafnyRef#sec-function-syntax).
+    ", range => new List<DafnyAction> {
+    OneAction("remove 'method'", range, "predicate", false),
+    OneAction("use 'ghost predicate'", range, "ghost predicate", false) // TODO - invalid if predicate is already ghost
```

Comment:
Remove this one. It's unlikely that someone wants, after sweating to write "function method", to have a ghost function instead.

Reply:
Removed